### PR TITLE
Noisy Networks for Exploration

### PR DESCRIPTION
**Pull request by markub3327**
# Description

NoisyNet is useful for users, that experiments with reinforcement learning algorithms (NoisyDQN or NoisyA3C). For more information, here is paper: [Noisy Networks for Exploration](https://arxiv.org/abs/1706.10295).

## Type of change

- [ ] Bug fix
- [ ] New Tutorial
- [ ] Updated or additional documentation
- [ ] Additional Testing
- [ ] New Activation and the changes conform to the [activation contribution guidelines](https://github.com/tensorflow/addons/blob/master/tensorflow_addons/activations/README.md#contribution-guidelines)
- [ ] New Callback and the changes conform to the [callback contribution guidelines](https://github.com/tensorflow/addons/blob/master/tensorflow_addons/callbacks/README.md#contribution-guidelines)
- [ ] New Image addition and the changes conform to the [image op contribution guidelines](https://github.com/tensorflow/addons/blob/master/tensorflow_addons/image/README.md#contribution-guidelines)
- [X] New Layer and the changes conform to the [layer contribution guidelines](https://github.com/tensorflow/addons/blob/master/tensorflow_addons/layers/README.md#contribution-guidelines)
- [ ] New Loss and the changes conform to the [loss contribution guidelines](https://github.com/tensorflow/addons/blob/master/tensorflow_addons/losses/README.md#contribution-guidelines)
- [ ] New Metric and the changes conform to the [metric contribution guidelines](https://github.com/tensorflow/addons/blob/master/tensorflow_addons/metrics/README.md#contribution-guidelines)
- [ ] New Optimizer and the changes conform to the [optimizer contribution guidelines](https://github.com/tensorflow/addons/blob/master/tensorflow_addons/optimizers/README.md#contribution-guidelines)
- [ ] New RNN Cell and the changes conform to the [rnn contribution guidelines](https://github.com/tensorflow/addons/blob/master/tensorflow_addons/rnn/README.md#contribution-guidelines)
- [ ] New Seq2seq addition and the changes conform to the [seq2seq contribution guidelines](https://github.com/tensorflow/addons/blob/master/tensorflow_addons/seq2seq/README.md#contribution-guidelines)
- [ ] New Text addition and the changes conform to the [text op contribution guidelines](https://github.com/tensorflow/addons/blob/master/tensorflow_addons/text/README.md#contribution-guidelines)

# Checklist:

- [ ] I've properly [formatted my code according to the guidelines](https://github.com/tensorflow/addons/blob/master/CONTRIBUTING.md#coding-style)
    - [ ] By running Black + Flake8
    - [ ] By running pre-commit hooks
- [ ] This PR addresses an already submitted issue for TensorFlow Addons
- [ ] I have made corresponding changes to the documentation
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] This PR contains modifications to C++ custom-ops

# How Has This Been Tested?

### Chart
W B Chart 29  8  2020, 11 06 32

NoisyNet is faster finding the optimal policy than Deep Q networks with epsilon-greedy.

### Code
[Noisy DQN MazeSolver](https://github.com/markub3327/DQN_MazeSolver)

### CODE REVIEW

**Comment by bhack** [comment]
I think this could fit better in [Tensorflow agents](https://github.com/tensorflow/agents/).
Let's check if @oars and @ebrevdo are interested to handle this in the agents repository.

**Comment by markub3327** [comment]
Hello,

I should open PR in https://github.com/tensorflow/agents/?

Thanks.

**Comment by ebrevdo** [comment]
I'll add sguada@ to this.  I see how this might go into tf_agents.keras_layers; though I think we would want to also see code that replicates the results in the NoisyNet paper.  But don't do anything yet, we might just ask that you keep it in your own repo for now.   Will chat w/ Sergio first.

**Comment by sguada** [comment]
You can take a look at and see if it would fit there. It would require extensive tests and clear api.
https://source.corp.google.com/piper///depot/google3/third_party/py/tf_agents/keras_layers/

**Comment by bhack** [comment]
> You can take a look at and see if it would fit there. It would require extensive tests and clear api.
> https://source.corp.google.com/piper///depot/google3/third_party/py/tf_agents/keras_layers/
> 

Is this an internal private URL?

**Comment by ebrevdo** [comment]
Yes; Sergio meant:
https://github.com/tensorflow/agents/tree/master/tf_agents/keras_layers

**Comment by bhack** [comment]
@ebrevdo @sguada It is ok for us to have this in your repository. What's next? Do you want that @markub3327 will open a DRAFT PR in your repo?

**Comment by tomerk** [comment]
Notes from ecosystem review: Would make sense in tf-agents if the tf-agents owners want it. We'll go ahead and ping @ebrevdo and @sguada again.

**Comment by tomerk** [comment]
Talked to @ebrevdo , it sounds like at this point in time tf-agents isn't looking to take on contributed layers that aren't directly used by their teams' experimentation. So, it'd probably be a better fit here?

**Comment by tomerk** [comment]
Doesn't look like I have label changing permissions to remove the ecosystem-review label here

**Comment by bhack** [comment]
> Doesn't look like I have label changing permissions to remove the ecosystem-review label here

done

**Comment by markub3327** [comment]
Sorry I have a new GitHub account and I don't have access to do changes in this PR. I made a new fork of 'tensorflow/addons' to my new profile with the same name @markub3327 as before.

**Comment by markub3327** [comment]
I create new PR with changes that do here: https://github.com/tensorflow/addons/pull/2338

**Comment by seanpmorgan** [comment]
Thank you for your contribution. We sincerely apologize for any delay in reviewing, but TensorFlow Addons is transitioning to a minimal maintenance and release mode. New features will not be added to this repository. For more information, please see our public messaging on this decision:
[TensorFlow Addons Wind Down](https://github.com/tensorflow/addons/issues/2807)

Please consider sending feature requests / contributions to other repositories in the TF community with a similar charters to TFA:
[Keras](https://github.com/keras-team/keras)
[Keras-CV](https://github.com/keras-team/keras-cv)
[Keras-NLP](https://github.com/keras-team/keras-nlp)
